packages/medusa/src/strategies/product/import.ts

```diff
@@ -0,0 +1,772 @@
+
+        if (!region) {
+          throw new MedusaError(
+            MedusaError.Types.INVALID_DATA,
+            `Trying to set a price for a region ${p.regionName} that doesn't exist`
+          )
+        }
+
+        record.region_id = region!.id
+      } else {
+        record.currency_code = p.currency_code
+      }
+
+      prices.push(record)
+    }
+
+    row["variant.prices"] = prices
+  }
+
+  /**
+   * A worker method called after a batch job has been created.
+   * The method parses a CSV file, generates sets of instructions
+   * for processing and stores these instructions to Redis.
+   *
+   * @param batchJobId . An id of a job that is being preprocessed.
+   */
+  async preProcessBatchJob(batchJobId: string): Promise<void> {
+    const batchJob = await this.batchJobService_.retrieve(batchJobId)
+
+    const csvFileKey = (batchJob.context as ImportJobContext).fileKey
+    const csvStream = await this.fileService_.getDownloadStream({
+      fileKey: csvFileKey,
+    })
+
+    const data = await this.csvParser_.parse(csvStream)
+    const results = await this.csvParser_.buildData(data)
+
+    const ops = await this.getImportInstructions(results)
+
+    await this.setImportDataToRedis(batchJobId, ops)
+
+    await this.batchJobService_.update(batchJobId, {
+      context: {
+        // number of update/create operations to execute
+        total: Object.keys(ops).reduce((acc, k) => acc + ops[k].length, 0),
+      },
+    })
+  }
+
+  /**
+   * The main processing method called after a batch job
+   * is ready/confirmed for processing.
+   *
+   * @param batchJobId - An id of a batch job that is being processed.
+   */
+  async processJob(batchJobId: string): Promise<void> {
+    return await this.atomicPhase_(async (transactionManager) => {
+      await this.createProducts(batchJobId, transactionManager)
+      await this.updateProducts(batchJobId, transactionManager)
+      await this.createVariants(batchJobId, transactionManager)
+      await this.updateVariants(batchJobId, transactionManager)
+
+      this.finalize_(batchJobId)
+    })
+  }
+
+  /**
+   * Method creates products using `ProductService` and parsed data from a CSV row.
+   *
+   * @param batchJobId - An id of the current batch job being processed.
+   * @param transactionManager - Transaction manager responsible for current batch import.
+   */
+  private async createProducts(
+    batchJobId: string,
+    transactionManager: EntityManager
+  ): Promise<void> {
+    const productOps = await this.getImportDataFromRedis(
+      batchJobId,
+      OperationType.ProductCreate
+    )
+
+    for (const productOp of productOps) {
+      try {
+        await this.productService_
+          .withTransaction(transactionManager)
+          .create(
+            transformProductData(productOp) as unknown as CreateProductInput
+          )
+      } catch (e) {
+        ProductImportStrategy.throwDescriptiveError(productOp)
+      }
+
+      this.updateProgress_(batchJobId)
```

Comment:
todo: Everything related to update the count on the class member should be moved to a proper place or passes as argument since we are in a singleton as mention in the comment at the property member level